[MRG+1] Trees as feature transformers

betatim opened the pull request:
This is an example using ensembles of trees to show how to use them to transform your samples into a high dimensional, sparse feature space and then train a linear model on that. In particular how to use the `apply` method of a `DecisionTree`. It came about from a discussion in #4488 and #4549. It is loosely based on @ogrisel's notebook here: http://nbviewer.ipython.org/github/ogrisel/notebooks/blob/master/sklearn_demos/Income%20classification.ipynb

In #4549 there is talk of showing the difference between a linear model and PCA. Not quite sure I got what you meant.

This could interest @amueller, @ogrisel and @vene.

betatim committed:
Example to demonstrate use of tree.apply() method

This example trains several tree based ensemble methods and uses
them to transform the data into a high dimensional, sparse space.
The trains a linear model on this new feature space. The idea is
taken from:

Practical Lessons from Predicting Clicks on Ads at Facebook Junfeng Pan,
He Xinran, Ou Jin, Tianbing XU, Bo Liu, Tao Xu, Yanxin Shi, Antoine
Atallah, Ralf Herbrich, Stuart Bowers, Joaquin Quiñonero Candela
International Workshop on Data Mining for Online Advertising (ADKDD)

https://www.facebook.com/publications/329190253909587/

vene commented on examples/ensemble/plot_feature_transformation.py:
```diff
@@ -0,0 +1,103 @@
+from sklearn.cross_validation import train_test_split
+from sklearn.metrics import roc_curve
+
+Nest = 10
```
Could you rename this `n_estimators`?

vene commented on examples/ensemble/plot_feature_transformation.py:
```diff
@@ -0,0 +1,103 @@
+
+Nest = 10
+X, y = make_classification(n_samples=80000)
+X_train,X_test, y_train,y_test = train_test_split(X, y, test_size=0.5)
```
pep8 spaces after commas (applicable below too)

betatim commented:
Content related: do you think as the example is now people can understand what it is meant to show (if they didn't already know)? I was thinking of putting a comment by the gradient boosting part to say "this is the interesting bit, using a tree's `apply` method".

Build related:
Any ideas why a lot of the gaussian process tests are failing in the py3.4 build?

In a previous build a test in `test_ridge.py` failed (https://ci.appveyor.com/project/sklearn-ci/scikit-learn/build/1.0.1287/job/wdqjboo3dvnw6lk4) which looks like a Heisenbug?

Should I file an issue for those or are they known to be strange?

amueller commented:
I think what I meant in #4549 was show a PCA on these vs on the original data? I'm not entirely following my wording, though.

amueller commented on examples/ensemble/plot_feature_transformation.py:
```diff
@@ -0,0 +1,103 @@
+"""Use trees to transform your features
```
You need to have a title like in the other plot examples which will be used in the example gallery.

amueller commented:
Thanks for working on this!

Can you post the plot here?
Also, it would be good to reference this in the user guide.

betatim commented:
![roc-curve](https://cloud.githubusercontent.com/assets/1448859/8943394/1b3ead0c-357b-11e5-81dc-6574a34274b4.png)

Undecided if zooming in on the top left (interesting) part is more helpful or confusing for others.

amueller commented:
the image looks white to me...

betatim commented:
Updated the plot, and added a reference in the ensemble section where `RandomTreesEmbedding` is discussed. I looked in "Dataset transformations" but couldn't find an existing topic where this would fit there. Somehow expected to see tree based transformations there as well.

If you can't quite remember the PCA comment, and I can't work it out either should we skip it/make a second example?

amueller commented:
looks good. Maybe it would be interesting to compare training the lr on the same training set vs a hold-out set? or at least mention that?

betatim commented:
You mean like this (just showing the RF part here but I changed it for all of the models):

```python
X, y = make_classification(n_samples=80000)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5)
X_train, X_train_lr, y_train, y_train_lr = train_test_split(X, y, test_size=0.5)

...
# Supervised transformation based on random forests
rf = RandomForestClassifier(max_depth=3, n_estimators=n_estimator)
rf_enc = OneHotEncoder()
rf_lm = LogisticRegression()
rf.fit(X_train, y_train)
rf_enc.fit(rf.apply(X_train))
rf_lm.fit(rf_enc.transform(rf.apply(X_train_lr)), y_train_lr)

...
```

![roc-curve](https://cloud.githubusercontent.com/assets/1448859/8967280/3cf0ebd4-3636-11e5-9082-725bf89077e0.png)

Hard to tell if it makes a difference. What is the idea behind using a different dataset for fitting the lr instead of using the same as for fitting the trees? We can change it or add it, but you'll have to provide the sentence to motivate it as I don't know enough ;)

amueller commented:
well the idea would be to be closer to "stacking". Intuitively training both on the same set should lead to crazy overfitting. but maybe not. let's keep it simple.

betatim commented:
Ok. I looked at 8.8 of ESL but I'll have to read it a few more times before it sinks in. Changed the example to use different subsets for fitting of the trees and the LR model.

amueller commented:
Well, I'm not entirely sure what the "industry standard" is.

glouppe commented on examples/ensemble/plot_feature_transformation.py:
```diff
@@ -0,0 +1,109 @@
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5)
+# It is important to train the ensemble of trees on a different subset
+# of the training data than the linear regression model to avoid overfitting
+X_train, X_train_lr, y_train, y_train_lr = train_test_split(X, y, test_size=0.5)
```
shouldnt you do `train_test_split(X_train, y_train, test_size=0.5)`?

betatim commented:
![roc-curve](https://cloud.githubusercontent.com/assets/1448859/9003431/3ded11aa-376e-11e5-8099-de294d9755c8.png)

Updated ROC curves. After discussing with @glouppe this morning I think it makes sense to split the training samples again. Updated the comment a bit to explain why/when this is important.

ndawe commented on examples/ensemble/plot_feature_transformation.py:
```diff
@@ -0,0 +1,113 @@
+
+Each sample goes through the decisions of each tree of the ensemble
+and ends up in one leaf per tree. The sample is encoded by setting
+feature values for these leafs to 1 and the other feature values to 0.
```
leafs -> leaves (?)

ndawe commented:
Showing inverse false positive rate on a log scale vs true positive rate will "enhance" the difference:

![figure_1](https://cloud.githubusercontent.com/assets/202816/9007088/59228aca-378e-11e5-80c3-f049bd7166a7.png)

Also changed n_samples to 10k, n_estimators to 20, and random seed to 10.

amueller commented:
With "enhanced" differences I would be a bit afraid of sending a message like "this is always better than that" when this is a very zoomed-in view of a single run on a particular synthetic dataset.

amueller commented:
(and currently it compares against a random forest using only part of the data, right?)

betatim commented:
I have a slight preference for the plain ROC curve. The original intent of the example was to show how to use the individual tree's `apply` method. (Though @glouppe was surprised that `GradientBoostedClassifier` does not have a public `apply`)

Right now it compares `RandomTreesEmbedding`, `RandomForestClassifier` and `GradientBoostedClassifier`. Each of these is used as input for a `LogisticRegression` which is then used to classify. For the last two it also shows the performance without the LR step. The trees and LR are trained on different subsets of the data. For the comparison without the LR we could retrain the tree ensembles on the whole training set...though that makes it feel like a lesson on whether or not you should use this technique, which was not the original intent.

larsmans commented:
Shouldn't this at least link to the `RandomTreesEmbedding` docs?

amueller commented:
Yeah, probably. I don't think we usually link from examples to docs. You can always click on the class to get to the api doc.

larsmans commented:
By link I meant at least mention it :) This seems to be doing the same thing that estimator does, except with more control over the type of trees.

betatim committed:
Explicitly mention RandomTreesEmbedding in the text

betatim commented:
It is so similar that we even use `RandomTreesEmbedding` in the example (RT+LR in the legend) :) What do you think of this modification?

glouppe commented:
I think the example is good enough to be merged. It is a nice demonstration of the `apply` method of tree-based methods. +1

amueller commented:
Thanks!

betatim commented:
Thanks! Now back to walking in the swiss alps: :mount_fuji:
